### PR TITLE
[release_conductor] Add framework cherrypicks substeps

### PR DESCRIPTION
**Pull request by Yugue**
Added `Apply framework cherrypicks` UI. Refactored `Apply engine cherrypicks` step widget to support framework as well. Both steps are similar and share the same widget.

Main Issues:
- [x] https://github.com/flutter/flutter/issues/94048
- [x] https://github.com/flutter/flutter/issues/94049

Side Issue:
- [x] Toggling between engine and framework uses hard-coded strings, convert to use enum instead.

Screen Capture with framework cherrypick conflicts:

Screen Shot 2021-11-22 at 1 31 29 PM

With no conflicts:

Screen Shot 2021-11-22 at 1 25 56 PM

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: release_dashboard/lib/logic/engine_or_framework.dart [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+// found in the LICENSE file.
+
+import '../enums/engine_or_framework.dart';
+
+/// Helper function that returns the string 'engine' or 'framework' based on the [engineOrFramework] enum.
+///
+/// Also supports returning first letter capitalized string.
+String engineOrFrameworkStr(EngineOrFramework engineOrFramework, [bool? capitalized = false]) {
+  if (capitalized == true) {
+    return engineOrFramework == EngineOrFramework.engine ? 'Engine' : 'Framework';
```
Comment (Yugue) [inline release_dashboard/lib/logic/engine_or_framework.dart:12]: Apparently, it is not straight-forward to make a logic to capitalize the first letter: https://stackoverflow.com/questions/29628989/how-to-capitalize-the-first-letter-of-a-string-in-dart

Returning the string hard-coded instead.

Comment (CaseyHillers) [reply release_dashboard/lib/logic/engine_or_framework.dart:12]: `'${engineOrFramework.toUpperCase()}${this.substring(1)}'`

As a side note, engineOrFramework should really just be named repository.

Comment (Yugue) [reply release_dashboard/lib/logic/engine_or_framework.dart]: Done, please take another look.

File: release_dashboard/lib/enums/repositories.dart [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/// A toggle enum to switch between engine and framework.
+enum Repositories {
```
Comment (CaseyHillers) [inline release_dashboard/lib/enums/repositories.dart:6]: Let's rebase once the enums PR has landed, then re-request my review (this mostly lgtm)

Comment (Yugue) [reply release_dashboard/lib/enums/repositories.dart]: Done, please take another look.

File: release_dashboard/lib/logic/repositories_str.dart [hunk 3]
```diff
@@ -0,0 +1,26 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import '../models/repositories.dart';
+
+const String kEngineStr = 'engine';
```
Comment (CaseyHillers) [inline release_dashboard/lib/logic/repositories_str.dart:7]: Avoid having types in the variable name. Would something like `kEngineRepo` make more sense?

File: release_dashboard/lib/logic/repositories_str.dart [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+const String kFrameworkStr = 'framework';
+
+/// Helper function that returns the string 'engine' or 'framework' based on the [Repositories] enum.
+///
+/// Also supports returning first letter capitalized string.
+String repositoriesStr(Repositories repository, [bool? capitalized = false]) {
```
Comment (CaseyHillers) [inline release_dashboard/lib/logic/repositories_str.dart:13]: Can we override `toString` in `Repositories` to handle this logic?

File: release_dashboard/lib/widgets/cherrypicks_substeps.dart [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+  @override
+  State<CherrypicksSubsteps> createState() => CherrypicksSubstepsState();
+
+  static const Map<CherrypicksSubstep, String> substepTitles = <CherrypicksSubstep, String>{
+    CherrypicksSubstep.verifyRelease: 'Verify the Release Number',
+    CherrypicksSubstep.applyCherrypicks: 'Apply cherrypicks and resolve conflicts',
+  };
+
+  static Map<pb.CherrypickState, String> cherrypickStates = <pb.CherrypickState, String>{
```
Comment (CaseyHillers) [inline release_dashboard/lib/widgets/cherrypicks_substeps.dart:44]: This is a good use case for [extension methods](https://dart.dev/guides/language/extension-methods). We could make an extension on `pb.CherrypickState` for `toString` and import that here.

File: release_dashboard/lib/logic/repositories_str.dart (hunk 4 above)
Comment (CaseyHillers) [reply release_dashboard/lib/logic/repositories_str.dart:13]: You can get temporarily unblocked by adding an extension method here, than when the upstream patch is available remove the extension method.

Comment (Yugue) [reply release_dashboard/lib/logic/repositories_str.dart:13]: Which `Repositories` you are referring to? The 'Repository` class in the core? Or the enum in the release dashboard called `Repositories`

Comment (Yugue) [reply release_dashboard/lib/logic/repositories_str.dart:13]: I just added an extension method here, please take another look.

File: release_dashboard/lib/widgets/cherrypicks_substeps.dart (hunk 5 above)
Comment (Yugue) [reply release_dashboard/lib/widgets/cherrypicks_substeps.dart:44]: Created an issue here: https://github.com/flutter/flutter/issues/94387, and added a todo here.

File: release_dashboard/lib/logic/repositories_str.dart (hunk 4 above)
Comment (Yugue) [reply release_dashboard/lib/logic/repositories_str.dart:13]: I also created a todo to have a method in the core

File: release_dashboard/lib/widgets/cherrypicks_substeps.dart (hunk 5 above)
Comment (CaseyHillers) [reply release_dashboard/lib/widgets/cherrypicks_substeps.dart:46]: Let's add the extension methods for this and `repositoryName` until those changes roll in

Comment (Yugue) [reply release_dashboard/lib/widgets/cherrypicks_substeps.dart:47]: Done, please take another look.

File: release_dashboard/lib/logic/cherrypick_state.dart [hunk 6]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:conductor_core/proto.dart' as pb;
+
+const Map<pb.CherrypickState, String> cherrypickStates = <pb.CherrypickState, String>{
```
Comment (CaseyHillers) [inline release_dashboard/lib/logic/cherrypick_state.dart:7]: This should be nested under cherryPickStateStr so it's no publically accessible (instead we should rely on the method)

File: release_dashboard/lib/logic/cherrypick_state.dart [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  /// {@tool snippet}
+  ///
+  /// ```dart
+  /// import "cherrypick_state.dart";
+  /// import 'package:conductor_core/proto.dart' as pb
+  ///
+  /// final String cherrypickStateStr = pb.CherrypickState.PENDING_WITH_CONFLICT.cherrypickStateStr();
+  /// ```
+  /// {@end-tool}
+  String cherrypickStateStr() {
```
Comment (CaseyHillers) [inline release_dashboard/lib/logic/cherrypick_state.dart:28]: `cherypickStateStr` seems repetitive, what about `String state => ...`?

File: release_dashboard/lib/logic/cherrypick_state.dart (hunk 6 above)
Comment (Yugue) [reply release_dashboard/lib/logic/cherrypick_state.dart:7]: Oops, forgot to remove this.

File: release_dashboard/lib/logic/cherrypick_state.dart (hunk 7 above)
Comment (Yugue) [reply release_dashboard/lib/logic/cherrypick_state.dart:28]: I changed the name to string.

**Review by CaseyHillers — APPROVED** [review]
flutter-lgtm

**Pull request merged**